### PR TITLE
uses varint encoding for vote-state lockout offsets

### PR DESCRIPTION
**Pull request by behzadnouri**
#### Problem
`VoteStateUpdate` has large serialized size.

#### Summary of Changes
The commit removes `CompactVoteStateUpdate` and instead reduces serialized
size of `VoteStateUpdate` using varint encoding for vote-state lockout
offsets.

### CODE REVIEW

File: sdk/program/src/vote/state/mod.rs [hunk 1]
```diff
@@ -789,6 +567,101 @@ impl VoteState {
...
+        },
+        serde::{Deserialize, Deserializer, Serialize, Serializer},
+    };
+
+    #[derive(Deserialize, Serialize, AbiExample)]
+    struct CompactLockout {
+        #[serde(with = "serde_varint")]
+        offset: Slot,
+        #[serde(with = "serde_varint")]
+        confirmation_count: u32,
```
Comment (AshwinSekar) [inline sdk/program/src/vote/state/mod.rs:586]: can be u8

**Review by AshwinSekar — COMMENTED** [review]
Since offset already has known bounds does this save us much vs using fixed? And is there a performance hit when serializing / deserializing varints?

Makes the code cleaner so I think it's a better solution, curious about the metrics.

**Comment by carllin** [comment]
Cool stuff, would be good to get a size comparison between this and the existing compactvotestate for a full tower (i.e. a vote with 32 slots)

File: sdk/program/src/vote/state/mod.rs (hunk 1 above)
Comment (behzadnouri) [reply sdk/program/src/vote/state/mod.rs]: done.

**Comment by behzadnouri** [comment]
> Since offset already has known bounds does this save us much vs using fixed? And is there a performance hit when serializing / deserializing varints?

yeah, might not save much. It depends on the actual value of offsets.
There is only one `lockouts` vector so we save the vector length encoding for the other two.
There is also no `root_to_first_vote_offset: u64`.
Also in cases where 21 bits are enough to encode the offset we save one byte compared to `lockouts_32`.

On the other hand there are instances we might need an extra byte due to most significant bit being used for varint encoding.
So at the end depends on the distribution of offsets.

The commit also removes `compact/uncompact` back-and-forth, and simplifies the code a lot. So it might be better in that regard even if the actual bytes savings are not significant.

File: sdk/program/src/vote/state/mod.rs (hunk 1 above)
Comment (AshwinSekar) [reply sdk/program/src/vote/state/mod.rs]: Was copied over from the original `Vote` which defined `confirmation_count` as `u32`. @carllin was there a specific reason for that?

**Comment by brennanwatt** [comment]
Collected some data comparing the encoding methods. Here's the average serialized size of vote state for each (sample size of 57k votes observed):
- VoteStateUpdate = 430B
- CompactVoteStateUpdate = 176B
- serde_compact_vote_state = 116B <-- Note my experiment added enum wrapper (extra 4B)

I was seeing very little variation for each. Very sparingly size would dip to the following minimum values
- VoteStateUpdate = 418B
- CompactVoteStateUpdate = 174B
- serde_compact_vote_state = 114B

Data can be found [here](https://docs.google.com/spreadsheets/d/1N1N_yHzyXntZcgc6KrSzaxWLz1o7Vc67FAoqF1h6Rus/edit?usp=sharing).

**Review by carllin — DISMISSED** [review]

**Review by AshwinSekar — APPROVED** [review]

**Pull request merged**